### PR TITLE
Port Microsoft.VisualBasic.FileSystem

### PR DESCRIPTION
**Pull request by cston**
Port `Microsoft.VisualBasic.FileSystem` from https://github.com/Microsoft/referencesource/tree/master/Microsoft.VisualBasic/runtime/msvbalib.

The changes from reference source were limited to comments, line breaks, and attributes. The implementation was not changed. Tests added for a subset of methods.

### CODE REVIEW

**Comment by cston** [comment]
@333fred, @agocke, @jaredpar, please review.
cc @danmosemsft, @KathleenDollard

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb [hunk 1]
```diff
@@ -1369,6 +1378,9 @@ Namespace Microsoft.VisualBasic
                         Throw VbMakeException(vbErrors.IllegalFuncCall)
                 End Select
             End If
+#Else
+            Throw New PlatformNotSupportedException()
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb]: You'll ideally want to have Unix tests that verify it throws PNSE (as oopposed to crashing somehow)

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IOUtils.vb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+Namespace Microsoft.VisualBasic.CompilerServices
+
+    <System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)> _
+    Class IOUtils
+        ' Prevent creation.
+        Private Sub New()
+        End Sub
+
+        <SecurityCritical()> _
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IOUtils.vb]: You can remove `Imports System.Security.Permissions` everywhere and `SecurityCritical` etc

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb [hunk 3]
```diff
@@ -0,0 +1,2994 @@
...
+                'Must come after all the Intrinsic types
+            ElseIf typ.IsValueType() Then
+                Return VT.Structure
+
+            Else
+                Return VT.Variant
+
+            End If
+        End Function
+
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb]: You may want to collapse all the multiple line breaks

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb [hunk 4]
```diff
@@ -0,0 +1,2994 @@
...
+                            m_bw.Write(LongType.FromObject(obj))
+                            m_position += 8
+
+                        Case VT.Single   '4 bytes
+                            LengthCheck(4)
+                            m_bw.Write(SingleType.FromObject(obj))
+                            m_position += 4
+
+                        Case VT.Error    '4 bytes
+                            Throw VbMakeException(vbErr
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb]: I guess we will never know what this means

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb [hunk 5]
```diff
@@ -3,15 +3,1690 @@
...
+                Return OpenModeTypes.Append
+            ElseIf (om = OpenMode.Binary) Then
+                Return OpenModeTypes.Binary
+            ElseIf (om = OpenMode.Random) Then
+                Return OpenModeTypes.Random
+            ElseIf CInt(om) = CInt(OpenModeTypes.Any) Then
+                Return OpenModeTypes.Any
+            End If
+
+            ' Fix FxCop violation. This is security-in-depth mean. This exception should never been hit.
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb]: remove comment ?

File: src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+            {
+                FileSystem.FileClose(0); // close all files
+            }
+            catch (Exception)
+            {
+            }
+            base.Dispose(disposing);
+        }
+
+        private static bool IsNotOSX() => !PlatformDetection.IsOSX;
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs]: Consider adding `IsNotOSX` to `PlatformDetection` isntead.

File: src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+            Assert.Throws<System.IO.IOException>(() => FileSystem.FileOpen(256, GetTestFilePath(), OpenMode.Append));
+        }
+
+        // Not tested:
+        //   public static void FileWidth(int FileNumber, int RecordWidth) { }
+
+        [Fact]
+        public void FreeFile()
+        {
+            int fileNumber = FileSystem.FreeFile();
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs]: I have never heard of this FreeFile concept. How odd!

**Comment by cston** [comment]
@333fred, @agocke, @jaredpar, please review, thanks.

File: src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+            var fileName = GetTestFilePath();
+            FileSystem.FileOpen(fileNumber, fileName, OpenMode.Output);
+            AssertThrows<ArgumentException>(() => FileSystem.Write(fileNumber, new object()));
+            FileSystem.FileClose(fileNumber);
+        }
+
+        // We cannot use XUnit.Assert.Throws<T>() for lambdas that rely on AssemblyData (such as
+        // file numbers) because AssemblyData instances are associated with the calling assembly, and
+        // in RELEASE builds, the calling assembly of a lambda invoked from XUnit.Assert.Throws<T>()
+        // is corelib rather than this assembly, so file numbers created outside the lambda will be invalid.
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic.Core/tests/FileSystemTests.cs]: @cston, I just tried replacing this with Assert.Throws, and all tests passed.  Can you elaborate on what's broken without this and why it's needed?

**Pull request merged**